### ISSUE TITLE
Problem with singleton with private ctor/dtor exposed via init in Python

### ISSUE THREAD

**Issue by cryos** [pybind/pybind11#947]
I am attempting to convert wrapping from Booost Python over to PyBind11, and I have been unable to translate some of my singleton wrapping code. I basically have:

```
class Singleton {
public:
  static Singleton& instance();
private:
  Singleton();
  ~Singleton();
};
```
Trying to wrap it with
```
  py::class_<Singleton,
             std::unique_ptr<Singleton, py::nodelete>>(m, "Singleton")
    .def("__init__", &Singleton::instance,
         py::return_value_policy::reference)
    .def("get_instance", &Singleton::instance,
         py::return_value_policy::reference);
```
This compiles, links, but when I try to run it the get_instance method works as expected, but the __init__ fails like so:
```
>>> f = module.Singleton()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
    1. () -> module.Singleton

Invoked with:
```
Is it possible to achieve this in PyBind11, or should I settle for telling people to use module.Singleton.get_instance() which works as expected? I would like to maintain as much of the Python API as I can with the switch.

**Comment by jagerman** [comment]
The issue with `__init__` is that it's too late to enforce a singleton on the Python side: the Python object has already been created.  The `incompatible arguments` error is happening because the implicit `self` argument which `__init__` gets called with isn't been accepted by your `instance()` method.

**Comment by cryos** [comment]
So this just isn't possible in the PyBind11 approach I guess? It is a shame, it worked well in Boost Python, but we could live without it if necessary, I was hoping I was missing something but nothing I tried worked. Do you want me to close this issue if this is as designed? Hopefully it will at least help others in the future trying to achieve this result, it might be useful to note in the docs that it is not possible.

**Comment by jagerman** [comment]
With PR #805, which I hope to merge soon, it should be possible to get two different Python objects that refer to the same singleton pointer.  They won't quite be singletons on the Python side (`a is b` won't be true) but in terms of anything on the c++ side they'll work as singletons.

**Comment by jagerman** [comment]
(This may well be what happens in Boost.Python as well, but I'm not familiar enough with it to say).

**Comment by cryos** [comment]
Thank you for the extra details, I am not familiar with their implementation but your patch looks like it would fit the bill nicely. Great to hear there is already some ongoing work, it is syntactic sugar for the Python side in this case, and I don't need a Python singleton - just seamless access to the same C++ singleton object.

**Comment by jagerman** [comment]
With #805 just merged, I think that your example will work if you change the `.def("__init__", &Singleton::instance, py::return_value_policy::reference)` to `.def(py::init(&Singleton::instance), py::return_value_policy::reference)`.  As I mentioned before, that'll give you multiple, distinct Python objects that reference the same C++ instance.

**Comment by dean0x7d** [comment]
Looks like this has been resolved with #805 so I'll close it. Feel free to reopen if there's anything more.
